Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/triangle_soup_snap_rounding.h

```diff
@@ -0,0 +1,550 @@
+
+namespace CGAL
+{
+
+namespace Polygon_mesh_processing
+{
+
+namespace internal
+{
+
+template <class NT> double double_ceil(Lazy_exact_nt< NT > x);
+template <class NT> double double_ceil(NT x);
+
+template <class NT>
+double double_ceil(Lazy_exact_nt< NT > x){
+  // If both sides are in the same ceil, return this ceil
+  double ceil_left=std::ceil(to_interval(x).first);
+  if(ceil_left==std::ceil(to_interval(x).second))
+    return ceil_left;
+  // If not refine interval by contracting the DAG and try again
+  x.exact();
+  ceil_left=std::ceil(to_interval(x).first);
+  if(ceil_left==std::ceil(to_interval(x).second))
+    return ceil_left;
+  // If not return the ceil of the exact value
+  return double_ceil( x.exact());
+};
+
+template <class NT>
+double double_ceil(NT x){
+  using FT = Fraction_traits<NT>;
+  if constexpr(std::is_same<typename FT::Is_fraction, Tag_true>::value){
+    // If NT is a fraction, the ceil value is the result of the euclidian division of the numerator and the denominator.
+    typename FT::Numerator_type num, r, e;
+    typename FT::Denominator_type denom;
+    typename FT::Decompose()(x,num,denom);
+    div_mod(num, denom, r, e);
+    if((r>=0) && e!=0) //If the result is positive, the ceil value is one above
+      return to_double(r+1);
+    return to_double(r);
+  } else {
+    // Return the ceil of the approximation
+    return std::ceil(to_double(x));
+  }
+};
+
+template <typename Range>
+class Indexes_range{
+  typedef std::remove_cv_t<typename std::iterator_traits<typename Range::iterator>::value_type> Value_type;
+public:
+  typedef typename Range::const_iterator const_iterator;
+  typedef typename Range::iterator iterator;
+
+  Indexes_range(){}
+  Indexes_range(std::initializer_list<size_t>  l): poly(l), m_id(0), modified(true){}
+  Indexes_range(const Indexes_range<Range> &ir):poly(ir.range()), m_id(ir.id()), modified(ir.was_modified()){}
+  Indexes_range(Range &p):poly(p), modified(true){}
+  Indexes_range(Range &p, size_t id):poly(p), m_id(id),modified(false){}
+
+  void operator=(std::initializer_list<size_t>  l)
+  {
+    poly=l;
+    modified=true;
+  }
+
+  inline size_t id() const { return m_id; }
+  inline void set_id(size_t id){ m_id=id; }
+  inline bool was_modified() const { return modified; }
+
+  inline size_t size() const{ return poly.size(); }
+
+  inline iterator begin(){ return poly.begin(); }
+  inline iterator end(){ return poly.end(); }
+  inline const_iterator begin() const{ return poly.begin(); }
+  inline const_iterator end() const{ return poly.end(); }
+
+  inline Value_type operator[](size_t i){ return poly[i]; }
+  inline const Value_type operator[](size_t i) const{ return poly[i]; }
+
+  inline void push_back(Value_type v){ poly.push_back(v);}
+  inline Range range(){ return poly;}
+  inline const Range range() const{ return poly;}
+private:
+  Range poly;
+  size_t m_id;
+  bool modified;
+};
+
+template <typename PolygonRange,typename Map>
+struct Wrapp_id_visitor
```

Comment:
If you inherit from the default visitor, you can only define function you'd like to override.